carbon-design-system/carbon-addons-iot-react
# fix(datetimepickerv2): flyout menu auto positioning

Closes #3618 

**Summary**

- expose `renderInPortal` prop so that user can set it to false for some scenarios
- expose `useAutoPositioning` prop to reposition flyout menu if offscreen



**Change List (commits, features, bugs, etc)**

- added `renderInPortal` prop in DateTimePickerV2, Card, TableCard, CardToolBar
- added `useAutoPositioning` prop in DateTimePickerV2
- fixed the menuOffSet when useAutoPositioning is enabled and in RTL. The passed-in `menuOffset` prop only works for the passed-in `direction` prop. If repositioned, the `direction` will be changed, `menuOffset` no long apply to the new `direction`
before:
![image](https://user-images.githubusercontent.com/24279794/198116800-47491546-b4a5-49be-8a79-b2ed06ecc69a.png)
after:
![image](https://user-images.githubusercontent.com/24279794/198116946-d77963ed-5585-4d3a-911b-3d80e9a1b0cd.png)

- fixed the menuOffSet.left. set it to 288px in RTL
Before,
![image](https://user-images.githubusercontent.com/24279794/198100853-18eb9ae6-a26e-47e9-9a9a-65eeafad71ac.png)

After:
![image](https://user-images.githubusercontent.com/24279794/198100989-78aecb35-e093-4c15-8aad-644c594ec28a.png)

- added tests for the scenarios when using `useAutoPositioning` and `renderInPortal`
- added a knob `useAutoPositioning` in single select and range select stories.

**Acceptance Test (how to verify the PR)**

- go to single select [story](https://deploy-preview-3608--carbon-addons-iot-react.netlify.app/?path=/story/2-watson-iot-experimental-%E2%98%A2%EF%B8%8F-datetimepickerv2--single-select)
- deselect padding on story-container 
- add margin-left to -1rem or so to make it offscreen on the left
- click to open flyout menu, make sure flyout menu are with in screen
![image](https://user-images.githubusercontent.com/24279794/198076205-ae0d12bd-e054-4889-ad46-f60685410851.png)

![image](https://user-images.githubusercontent.com/24279794/198076410-076e76d4-a6d3-4a0d-bc5d-f67686cb8cd4.png)



- go to range select or single select [story](https://deploy-preview-3608--carbon-addons-iot-react.netlify.app/?path=/story/2-watson-iot-experimental-%E2%98%A2%EF%B8%8F-datetimepickerv2--selected-absolute-with-new-time-spinner)
- set renderInPortal to false, we should see `flyout-tooltip` is grandchild of `date-time-picker-datepicker-flyout-container`
![image](https://user-images.githubusercontent.com/24279794/196543175-86271623-a191-4044-9fef-5d870a51f993.png)

- set renderInPortal to true,  `flyout-tooltip` is independent element outside of `date-time-picker-datepicker-flyout-container`
![image](https://user-images.githubusercontent.com/24279794/196543557-e93d0471-b375-4d0f-8f79-1a05334eeda7.png)




**Things to look for during review**
